spec/offline_sort/offline_sort_spec.rb

```diff
@@ -1,113 +1,138 @@
+    let(:offline_sorter_instance) do
+      described_class.new(enumerable, chunk_size: entries_per_chunk, &sort)
     end
 
-    it "writes out to disk" do
-      expect(Tempfile).to receive(:open).at_least(:once).and_call_original
-      subject
+    let(:entries_per_chunk) { 900 }
+    let(:count) { 10000 }
+    let(:enumerable) { arrays }
+    let(:sort) { array_sort }
+    let(:array_sort) { Proc.new { |arr| arr[2] } }
+    let(:arrays) do
+      count.times.map do |index|
+        [SecureRandom.hex, index, SecureRandom.hex]
+      end
     end
 
-    shared_examples "produces a sorted result" do
-      it "produces the same sorted result as an in-memory sort" do
-        sorted = subject.to_a
+    shared_examples "a correct offline sort" do
+      let(:unsorted) { enumerable.dup }
+
+      it "writes out to disk" do
+        expect(Tempfile).to receive(:open).at_least(:once).and_call_original
+        subject
+      end
+
+      shared_examples "produces a sorted result" do
+        it "produces the same sorted result as an in-memory sort" do
+          expect(unsorted).to match_array(enumerable)
+          expect do
+            last = nil
+            entry_count = 0
+            subject.each do |entry|
+              if last.nil?
+                last = entry
+                entry_count += 1
+                next
+              end
+
+              unless (sort.call(last) <=> sort.call(entry)) == -1
+                raise "Out of order at line #{entry_count}"
+              end
 
-        expect(unsorted).to match_array(enumerable)
-        expect do
-          last = nil
-          entry_count = 0
-          sorted.each do |entry|
-            if last.nil?
               last = entry
               entry_count += 1
-              next
             end
+          end.not_to raise_error
+          expect(subject).to match_array(enumerable.sort_by(&sort))
+        end
+
+        context "closing tempfiles" do
+          it "closes all tempfiles" do
+            close_count = 0
+            allow_any_instance_of(Tempfile).
+              to receive(:close) { close_count += 1 }
+
+            expected_number_of_tempfiles =
+              (count.to_f / entries_per_chunk).ceil
 
-            unless (sort.call(last) <=> sort.call(entry)) == -1
-              raise "Out of order at line #{entry_count}"
+            # The case where we don't write to disk because there's only
+            # one chunk
+            if expected_number_of_tempfiles == 1
+              expected_number_of_tempfiles = 0
             end
 
-            last = entry
-            entry_count += 1
+            subject
+            expect(close_count).to eq(expected_number_of_tempfiles)
           end
-        end.not_to raise_error
-        expect(sorted).to match_array(enumerable.sort_by(&sort))
+        end
+
+        context "when sorted twice" do
+          it "produces the same result both times" do
```

Comment:
Could/should this have some test that we don't actually run the underlying sort multiple times? One way of doing this would be to have the `sort` proc count the number of times it's called, so that you can write an assertion around the number of times that work is done?